Notes/BeetlebergsDesign.md

```diff
@@ -0,0 +1,377 @@
+
+All the available sprite vector attributes are diagrammed here:
+
+![The four attributes describing the sprite: change_x, change_y, angle, and change_angle.](BeetleSpriteVectors.png)
+
+So the proposal is that we keep track of two scalar floats of speed, `forward_speed` and `sideways_speed`. Note that
+these are scalar, not vector values, but that's OK because so are the sprite attributes. Essentially, we will convert
+these speed values by finding the global x and y components of the vectors they describe based on `angle`. It might be
+useful to show this process on a diagram:
+
+![A diagram of our calculations of the global x and y components of our forward and sideways vectors.](BeetleCalculations.png)
+
+While `forward_speed` and `sideways_speed` are scalars, the vectors they describe are drawn in green. We will have to
+calculate the purple vectors that are the x and y components. Luckily, we can use basic trigonometry here. Take notice
+of θ, which is 90 - `angle`. This makes sin(θ) equal to the length of the y-component of the forward vector divided
+the length of the length of the forwards vector. We can reason likewise for the x-component and cos(θ). Putting this
+together, and realizing that if θ is 90 - `angle`, then it is the other angle in the triangle, and thus we can just use
+`angle` directly:
+
+> `forward_x` = sin(`angle`) \* `forward_speed`  
+> `forward_y` = cos(`angle`) \* `forward speed`  
+> `sideways_x` = cos(`angle`) \* `sideways_speed`  
+> `sideways_y` = sin(`angle`) \* `sideways_speed`  
+> `change_x` = `forward_x` + `sideways_x`  
+> `change_y` = `forward_y` + `sideways_y`  
+
+Thus this is some pseudo-code for handling movement.
+
+### Beetle Stats
+
+Each beetle has a collection of stats that will be represented as attributes on the base `Beetle` class:
+
+* `max_forward_speed`: How fast the beetle can move along the facing axis (possibly separate `max_backward_speed`?)
+* `max_sideways_speed`: How fast, if at all, the beetle can move along the left-right axis perpendicular to the facing
+  axis. A beetle will not have this by default, but gain it through a trait.
+* `accuracy`: How good the beetle is at aiming projectiles or perhaps aiming itself so basic projectiles hit (not sure
+  exactly how to calculate/apply this, to be honest).
+* `max_rotation_speed`: How fast the beetle can turn.
+* `hit_points`: How much damage the beetle can take before being eliminated.
+* `toughness`: How much, if any, damage is negated before it is applied to `hit_points`. In theory, this could also be a
+  malus where a beetle takes _extra_ damage! A beetle will only have this by gaining it through a trait.
+* `size`: How large the beetle is, increasing its drawn size and thus its hitbox.
+* `awareness`: The beetle's ability to notice the location of enemies and respond to incoming projectiles (these could
+  separate stats, like `dodge_percentage` and `enemy_awareness`).
```

Comment:
It occurs to me on re-reading this that the projectile counts are assuming the beetle is only watching enemy projectiles. I think this is fine, but I am open to models where it accounts for all projectiles regardless of source. But since we're not doing friendly fire, I think enemy-only makes the most sense for now.

Reply:
I like this idea, it reminds me of Miu Miu's stand ability in part 6, which I think then we could exactly do your last idea where awareness allows them to hold more things in mind. We should also think about making the preferences here customizable, that will be used to discard things it can't remain aware of.

We could eventually let the player manipulate their preferences, but at the least for our ability to getting the beetles to behave with the right level of intelligence we want, we're going to need it as devs.

Reply:
OK, I directly stole this diagram and made some slight changes to the scenario but felt it was a good example to describe how this works. I also described what we agreed upon regarding awareness.